lib/core_to_mucore.ml

```diff
@@ -1091,13 +1146,19 @@ let make_fun_with_spec_args f_i loc env args (accesses, requires) =
       (*   let here = Locations.other __LOC__ in *)
       (*   (LTranslate.T (IT.good_ (ct, IT.sym_ (pure_arg, bt, here)) here), info) *)
       (* in *)
-      let@ at = aux (* good_lc :: *) good_lcs env st rest in
+      let@ at = aux (* good_lc :: *) good_lcs env st (rest, ghost_args) in
       return (Mu.mComputational ((pure_arg, bt), (loc, None)) at)
-    | [] ->
-      let@ lat = make_largs_with_accesses f_i env st (accesses, requires) in
+    | [], (arg, cnbt) :: rest ->
+      let sbt = Compile.base_type env cnbt in
+      let bt = SBT.proj sbt in
+      let env = Translate.add_logical arg sbt env in
+      let@ at = aux good_lcs env st ([], rest) in
+      return (Mu.Ghost ((arg, bt), (loc, None), at))
+    | [], [] ->
+      let@ lat = make_largs_with_accesses f_i env st (accesses, snd requires) in
       return (Mu.L (Mu.mConstraints (List.rev good_lcs) lat))
   in
-  aux [] env Translate.C_vars.init args
+  aux [] env Translate.C_vars.init (args, fst requires)
```

Comment:
Would this be clearer as two recursive functions, rather than one with `ghost_args` threaded through?